xdsl/interpreters/eqsat_pdl_interp.py

```diff
@@ -97,3 +113,56 @@ def run_getresults(
                     " that are used by a single EClassOp each."
                 )
         return (results,)
+
+    @impl(pdl_interp.GetDefiningOpOp)
+    def run_getdefiningop(
+        self,
+        interpreter: Interpreter,
+        op: pdl_interp.GetDefiningOpOp,
+        args: tuple[Any, ...],
+    ) -> tuple[Any, ...]:
+        defining_op = cast(
+            None | Operation,
+            PDLInterpFunctions.run_get_defining_op(self, interpreter, op, args).values[
+                0
+            ],
+        )
+        if isinstance(eclass_op := defining_op, eqsat.EClassOp):
+            if not self.visited:
+                if op != self.backtrack_stack[-1].gdo_op:
+                    raise InterpretationError(
```

Comment:
can you please add a comment explaining how this condition is caused by multiple `get_defining_op`s in a block?

Reply:
Actually more comments below would be great, I'm not 100% sure I understand the mechanism here